src/systems/filecoin_markets/storage_market/storage_market_actor_state.go

```diff
@@ -0,0 +1,288 @@
+	// TODO
+	var dealID deal.DealID
+	return dealID
+}
+
+func (st *StorageMarketActorState_I) _isBalanceAvailable(a addr.Address, amount actor.TokenAmount) bool {
+	bal := st.Balances()[a]
+	return bal.Available() >= amount
+}
+
+func (st *StorageMarketActorState_I) _assertValidClienSignature(rt Runtime, dealP deal.StorageDealProposal) {
+	// TODO: verify if we need to check provider signature
+	// or it is implicit in the message
+
+	// Optimization: make this a batch verification
+	panic("TODO")
+}
+
+func (st *StorageMarketActorState_I) _assertDealStartAfterCurrEpoch(rt Runtime, p deal.StorageDealProposal) {
+
+	currEpoch := rt.CurrEpoch()
+
+	// deal has started before or in current epoch
+	if p.StartEpoch() <= currEpoch {
+		rt.Abort("sma._assertDealStartAfterCurrEpoch: deal started before or in CurrEpoch.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertDealNotYetExpired(rt Runtime, p deal.StorageDealProposal) {
+	currEpoch := rt.CurrEpoch()
+
+	if p.EndEpoch() <= currEpoch {
+		rt.Abort("st._assertDealNotYetExpired: deal has expired.")
+	}
+}
+
+func (st *StorageMarketActorState_I) _assertValidDealTimingAtPublish(rt Runtime, p deal.StorageDealProposal) {
+
+	// TODO: verify deal did not expire when it is signed
+
+	st._assertDealStartAfterCurrEpoch(rt, p)
+
+	// deal ends before it starts
+	if p.EndEpoch() <= p.StartEpoch() {
+		rt.Abort("sma._assertValidDealTimingAtPublish: deal ends before it starts.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertValidDealMinimum(rt Runtime, p deal.StorageDealProposal) {
+
+	// minimum deal duration
+	if p.Duration() < deal.MIN_DEAL_DURATION {
+		rt.Abort("sma._assertValidDealMinimum: deal duration shorter than minimum.")
+	}
+
+	if p.StoragePricePerEpoch() <= deal.MIN_DEAL_PRICE {
+		rt.Abort("sma._assertValidDealMinimum: storage price less than minimum.")
+	}
+
+	// verify StorageDealCollateral match requirements for MinimumStorageDealCollateral
+	if p.ProviderCollateralPerEpoch() < deal.MIN_PROVIDER_DEAL_COLLATERAL_PER_EPOCH ||
+		p.ClientCollateralPerEpoch() < deal.MIN_CLIENT_DEAL_COLLATERAL_PER_EPOCH {
+		rt.Abort("sma._assertValidDealMinimum: deal collaterals less than minimum.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertSufficientBalanceAvailForDeal(rt Runtime, p deal.StorageDealProposal) {
+
+	// verify client and provider has sufficient balance
+	isClientBalAvailable := st._isBalanceAvailable(p.Client(), p.ClientBalanceRequirement())
+	isProviderBalAvailable := st._isBalanceAvailable(p.Provider(), p.ProviderBalanceRequirement())
+
+	if !isClientBalAvailable || !isProviderBalAvailable {
+		rt.Abort("sma._validateNewStorageDeal: client or provider insufficient balance.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertDealExpireAfterMaxProveCommitWindow(rt Runtime, dealP deal.StorageDealProposal) {
+
+	currEpoch := rt.CurrEpoch()
+	dealExpiration := dealP.EndEpoch()
+
+	if dealExpiration <= (currEpoch + sector.MAX_PROVE_COMMIT_SECTOR_EPOCH) {
```

Comment:
should this be <?

at least X sounds inclusive of X

Reply:
I think it should still be <= since it will throw if this is true.

Reply:
take 
`currEpoch == 10`
`MAX_PROVE == 15`
`EndEpoch == 25`

Then in your case it would be triggered whereas in mine it would not, even though EndEpoch is MAX_PROVE away from currEpoch. So we should fix language accordingly: it is either
- at least MAX_PROVE (<), or,
- more than MAX_PROVE (<=)

Reply:
I am not sure where the language is coming from, the method is called `_assertDealExpireAfterMaxProveCommit`, in the example you mentioned, the deal is expiring on MaxProveCommit and hence it is not after ProveCommit and hence it will throw.